In [1]:
import warnings
warnings.simplefilter('ignore')

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

# Load Dataset

In [2]:
# read raw data
data_2011 = pd.read_csv('./data/2011.csv')
data_2013 = pd.read_csv('./data/2013.csv')
data_2014 = pd.read_csv('./data/2014.csv')
data_2015 = pd.read_csv('./data/2015.csv')
data_2016_1 = pd.read_csv('./data/2016(1).csv')
data_2016_2 = pd.read_csv('./data/2016(2).csv')
data_2016_3 = pd.read_csv('./data/2016(3).csv')
data_2016_4 = pd.read_csv('./data/2016(4).csv')
data_2017_1 = pd.read_csv('./data/2017(1).csv')
data_2017_2 = pd.read_csv('./data/2017(2).csv')
data_2017_3 = pd.read_csv('./data/2017(3).csv')
data_2017_4 = pd.read_csv('./data/2017(4).csv')
data_2018 = pd.read_csv('./data/2018(1).csv')

# merge data from 2016 and 2017
data_2016 = pd.concat(objs=[data_2016_1, data_2016_2, data_2016_3, data_2016_4], axis=0)
data_2017 = pd.concat(objs=[data_2017_1, data_2017_2, data_2017_3, data_2017_4], axis=0)

In [14]:
# basic information
print('Year 2011:\t', data_2011.shape)
print('Year 2013:\t', data_2013.shape)
print('Year 2014:\t', data_2014.shape)
print('Year 2015:\t', data_2015.shape)
print('Year 2016:\t', data_2016.shape)
print('Year 2017:\t', data_2017.shape)
print('Year 2018:\t', data_2018.shape)

Year 2011:	 (42538, 143)
Year 2013:	 (188181, 143)
Year 2014:	 (235629, 143)
Year 2015:	 (421095, 143)
Year 2016:	 (434407, 143)
Year 2017:	 (443579, 143)
Year 2018:	 (107864, 143)


In [15]:
data_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42538 entries, 0 to 42537
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(114), object(29)
memory usage: 46.4+ MB


In [16]:
data_2013.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188181 entries, 0 to 188180
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(92), int64(15), object(36)
memory usage: 205.3+ MB


In [17]:
data_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235629 entries, 0 to 235628
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(66), int64(41), object(36)
memory usage: 257.1+ MB


In [18]:
data_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421095 entries, 0 to 421094
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(66), int64(40), object(37)
memory usage: 459.4+ MB


In [19]:
data_2016.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 434407 entries, 0 to 103545
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(66), int64(40), object(37)
memory usage: 477.3+ MB


In [20]:
data_2017.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 443579 entries, 0 to 118647
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(55), int64(51), object(37)
memory usage: 487.3+ MB


In [21]:
data_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107864 entries, 0 to 107863
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(59), int64(54), object(30)
memory usage: 117.7+ MB


# Data Concatenation

After observation in depth, we notice that early date, especially the data before 2011, have some difference with the most recent data, such as the data in 2017 and 2018. Since the data before 2013 is only a small part of all data (~$12\%$), in this project, we will only use the data from 2014 until 2018.

In [143]:
objs = [data_2014, data_2015, data_2016_1, data_2016_2, data_2016_3, data_2016_4, 
        data_2017_1, data_2017_2, data_2017_3, data_2017_4, data_2018]
data = pd.concat(objs=objs, axis=0)
data.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0      10400        10400          10400.0   36 months    6.99%       321.08   
1      15000        15000          15000.0   60 months   12.39%       336.64   
2       9600         9600           9600.0   36 months   13.66%       326.53   
3       7650         7650           7650.0   36 months   13.66%       260.20   
4      21425        21425          21425.0   60 months   15.59%       516.36   

  grade sub_grade                        emp_title emp_length       ...        \
0     A        A3   Truck Driver Delivery Personel    8 years       ...         
1     C        C1                       MANAGEMENT  10+ years       ...         
2     C        C3                 Admin Specialist  10+ years       ...         
3     C        C3             Technical Specialist   < 1 year       ...         
4     D        D1  Programming Analysis Supervisor    6 years       ...         

  hardship_payoff_balance_amount  hardship_last_payment_amount  \
0                            NaN                           NaN   
1                            NaN                           NaN   
2                            NaN                           NaN   
3                            NaN                           NaN   
4                            NaN                           NaN   

  disbursement_method debt_settlement_flag debt_settlement_flag_date  \
0                Cash                    N                       NaN   
1                Cash                    N                       NaN   
2                Cash                    N                       NaN   
3                Cash                    N                       NaN   
4                Cash                    N                       NaN   

  settlement_status  settlement_date settlement_amount settlement_percentage  \
0               NaN              NaN               NaN                   NaN   
1               NaN              NaN               NaN                   NaN   
2               NaN              NaN               NaN                   NaN   
3               NaN              NaN               NaN                   NaN   
4               NaN              NaN               NaN                   NaN   

  settlement_term  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  

[5 rows x 143 columns]

In [144]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1642574 entries, 0 to 107863
Columns: 143 entries, loan_amnt to settlement_term
dtypes: float64(67), int64(38), object(38)
memory usage: 1.8+ GB


# Data Cleaning

In [163]:
# missing values
pd.options.display.max_rows = 150
missing = data.isnull().sum()
missing_ratio = missing / len(data)
missing_ratio = missing_ratio.reset_index()

# check the feature that has missing ration >= 0.2
missing_ratio = missing_ratio.rename(columns={'index': 'feature', 0: 'missing ratio'})
missing_ratio = missing_ratio.sort_values(by='missing ratio', ascending=False)
missing_ratio[missing_ratio['missing ratio'] >= 0.2]

feature  missing ratio
16                                          url       1.000000
132  orig_projected_additional_accrued_interest       0.996483
126                         hardship_start_date       0.995612
127                           hardship_end_date       0.995612
128                     payment_plan_start_date       0.995612
129                             hardship_length       0.995612
130                                hardship_dpd       0.995612
131                        hardship_loan_status       0.995612
133              hardship_payoff_balance_amount       0.995612
134                hardship_last_payment_amount       0.995612
125                             hardship_amount       0.995612
124                               deferral_term       0.995612
123                             hardship_status       0.995612
122                             hardship_reason       0.995612
121                               hardship_type       0.995612
17                                         desc       0.990657
119         sec_app_mths_since_last_major_derog       0.988438
142                             settlement_term       0.988106
137                   debt_settlement_flag_date       0.988106
139                             settlement_date       0.988106
140                           settlement_amount       0.988106
141                       settlement_percentage       0.988106
138                           settlement_status       0.988106
114                          sec_app_revol_util       0.966871
109                             revol_bal_joint       0.966293
115                         sec_app_open_act_il       0.966293
110                    sec_app_earliest_cr_line       0.966293
111                      sec_app_inq_last_6mths       0.966293
113                            sec_app_open_acc       0.966293
112                            sec_app_mort_acc       0.966293
116                       sec_app_num_rev_accts       0.966293
117            sec_app_chargeoff_within_12_mths       0.966293
118          sec_app_collections_12_mths_ex_med       0.966293
53                    verification_status_joint       0.958762
52                                    dti_joint       0.958570
51                             annual_inc_joint       0.958567
27                       mths_since_last_record       0.823255
84                     mths_since_recent_bc_dlq       0.752743
48                  mths_since_last_major_derog       0.721055
86               mths_since_recent_revol_delinq       0.650519
26                       mths_since_last_delinq       0.491939
45                                 next_pymnt_d       0.485112
63                                      il_util       0.471999
61                           mths_since_rcnt_il       0.404905
67                                     all_util       0.386912
57                                  open_acc_6m       0.386840
70                                  total_cu_tl       0.386840
71                                 inq_last_12m       0.386840
58                                  open_act_il       0.386840
59                                  open_il_12m       0.386840
66                                   max_bal_bc       0.386840
65                                  open_rv_24m       0.386840
64                                  open_rv_12m       0.386840
62                                 total_bal_il       0.386840
60                                  open_il_24m       0.386840
69                                       inq_fi       0.386840

In [169]:
# it's safe to remove the features with missing ratio >= 0.3
columns = ['url', 'orig_projected_additional_accrued_interest', 'hardship_start_date', 
           'hardship_end_date', 'payment_plan_start_date', 'hardship_length', 
           'hardship_dpd', 'hardship_loan_status', 'hardship_payoff_balance_amount', 
           'hardship_last_payment_amount', 'hardship_amount', 'deferral_term', 
           'hardship_status', 'hardship_reason', 'hardship_type', 'desc', 
           'sec_app_mths_since_last_major_derog', 'settlement_term', 
           'debt_settlement_flag_date', 'settlement_date', 'settlement_amount', 
           'settlement_percentage', 'settlement_status', 'sec_app_revol_util', 
           'revol_bal_joint', 'sec_app_open_act_il', 'sec_app_earliest_cr_line', 
           'sec_app_inq_last_6mths', 'sec_app_open_acc', 'sec_app_mort_acc', 
           'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths', 
           'sec_app_collections_12_mths_ex_med', 'verification_status_joint',
           'dti_joint', 'annual_inc_joint', 'mths_since_last_record', 
           'mths_since_recent_bc_dlq', 'mths_since_last_major_derog', 
           'mths_since_recent_revol_delinq', 'mths_since_last_delinq', 'next_pymnt_d', 
           'il_util', 'mths_since_rcnt_il', 'all_util', 'open_acc_6m', 'total_cu_tl', 
           'inq_last_12m', 'open_act_il', 'open_il_12m', 'max_bal_bc', 'open_rv_24m', 
           'open_rv_12m', 'total_bal_il', 'open_il_24m', 'inq_fi']

In [182]:
# drop features with too many missing values
data = data.drop(labels=columns, axis=1)

data.to_csv('./data/data_2014_2018.csv', index=False)

data.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0      10400        10400          10400.0   36 months    6.99%       321.08   
1      15000        15000          15000.0   60 months   12.39%       336.64   
2       9600         9600           9600.0   36 months   13.66%       326.53   
3       7650         7650           7650.0   36 months   13.66%       260.20   
4      21425        21425          21425.0   60 months   15.59%       516.36   

  grade sub_grade                        emp_title emp_length  \
0     A        A3   Truck Driver Delivery Personel    8 years   
1     C        C1                       MANAGEMENT  10+ years   
2     C        C3                 Admin Specialist  10+ years   
3     C        C3             Technical Specialist   < 1 year   
4     D        D1  Programming Analysis Supervisor    6 years   

          ...          percent_bc_gt_75  pub_rec_bankruptcies tax_liens  \
0         ...                      14.3                     0         0   
1         ...                       0.0                     0         0   
2         ...                      60.0                     0         0   
3         ...                     100.0                     0         0   
4         ...                     100.0                     0         0   

  tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit  \
0          179407             15030          13000                      11325   
1          196500            149140          10000                      12000   
2           52490             38566          21100                      24890   
3           82331             64426           4900                      64031   
4           57073             42315          15000                      35573   

  hardship_flag disbursement_method debt_settlement_flag  
0             N                Cash                    N  
1             N                Cash                    N  
2             N                Cash                    N  
3             N                Cash                    N  
4             N                Cash                    N  

[5 rows x 87 columns]

In [183]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1642574 entries, 0 to 107863
Data columns (total 87 columns):
loan_amnt                     1642574 non-null int64
funded_amnt                   1642574 non-null int64
funded_amnt_inv               1642574 non-null float64
term                          1642574 non-null object
int_rate                      1642574 non-null object
installment                   1642574 non-null float64
grade                         1642574 non-null object
sub_grade                     1642574 non-null object
emp_title                     1535014 non-null object
emp_length                    1537225 non-null object
home_ownership                1642574 non-null object
annual_inc                    1642574 non-null float64
verification_status           1642574 non-null object
issue_d                       1642574 non-null object
loan_status                   1642574 non-null object
pymnt_plan                    1642574 non-null object
purpose                

Finally we got the data used for modeling after some preliminary processing and cleaning. The cleaned dataset contains 1,642,574 records with 87 features. 

Further cleaning and processing are presented in following notebooks.